inference-engine/src/vpu/graph_transformer/src/model/model.cpp

```diff
@@ -1511,6 +1591,35 @@ DataToDataAllocation ModelObj::connectDataWithDataImpl(
     return edge;
 }
 
+namespace {
+
+bool isStageDependencyNeeded(
+        const Stage& dependentStage,
+        const Data& dependency) {
+    const auto& dependencyProducer = dependency->producer();
+
+    if (dependencyProducer == nullptr) {
+        return false;
+    }
+
+    // Check one level above, it covers current cases while checking all the levels might be computationally expensive
+    for (const auto& prevStage : dependencyProducer->prevStages()) {
+        if (prevStage == dependentStage) {
+            return false;
+        }
+    }
+
+    for (const auto& prevStage : dependentStage->prevStages()) {
+        if (prevStage == dependencyProducer) {
+            return false;
+        }
+    }
+
+    return dependentStage == dependencyProducer ? false : true;
```

Comment:
Isn't it better to check this condition at the beginning of this method? To avoid checking all previous stages in such cases?

Reply:
Agree. Fixed.